lib/cli/src/Cardano/CLI.hs

```diff
@@ -296,6 +331,202 @@ runCli = join . customExecParser preferences
+         formatEnglishEnumeration = formatEnglishEnumerationRev . reverse
+         formatEnglishEnumerationRev [ult, penult]
+            = penult ++ " or " ++ ult
+         formatEnglishEnumerationRev (ult:penult:revBeginning)
+            = intercalate ", " (reverse revBeginning)
+                ++ ", "
+                ++ penult
+                ++ " or "
+                ++ ult
+         formatEnglishEnumerationRev xs = intercalate ", " (reverse xs)
+
+parseSeedWords :: Parser [Text]
+parseSeedWords = some (argument str (metavar "MNEMONIC_WORDS..."))
+
+-- | Record with mnemonic and key derivation funcionality — /without/ any type
+-- parameters related to scheme.
+--
+-- This means that we can have a value for byron, a value for icarus, both with
+-- the same type @CliKeyScheme@.
+--
+-- @CliKeyScheme@ is on the other hand parameterized over @key@ and @m@.
+-- @hoistKeyScheme@ is provided for mapping over @m@. @mapKey@ is provided for
+-- mapping over @key@.
+--
+-- This way we can test mapping @XPrv@s to @Text@ as pure code in
+-- @Either String@, rather than @IO@.
+data CliKeyScheme key m = CliKeyScheme
+    { allowedWordLengths :: [Int]
+    , mnemonicToRootKey :: [Text] -> m key
+    }
+
+-- | Change the underlying monad of a @CliKeyScheme@.
+hoistKeyScheme
+    :: (forall a. m1 a -> m2 a)
+    -> CliKeyScheme key m1
+    -> CliKeyScheme key m2
+hoistKeyScheme eta s = CliKeyScheme
+    { allowedWordLengths = allowedWordLengths s
+    , mnemonicToRootKey = eta . mnemonicToRootKey s
+    }
+
+xPrvToHexText :: XPrv -> Text
+xPrvToHexText = T.pack . B8.unpack . hex . unXPrvStripPub
+
+unsafeHexTextToXPrv :: Text -> XPrv
+unsafeHexTextToXPrv t = either error Prelude.id $
+    xPrvFromStrippedPubXPrv
+    . unsafeFromHex
+    . B8.pack
+    . T.unpack
+    $ t
+
+-- | Map over the key type of a CliKeyScheme.
+--
+-- Can be used with e.g @xPrvToHexText@.
+mapKey
+    :: Functor m
+    => (key1 -> key2)
+    -> CliKeyScheme key1 m
+    -> CliKeyScheme key2 m
+mapKey f s = CliKeyScheme
+    { allowedWordLengths = allowedWordLengths s
+    , mnemonicToRootKey = fmap f . mnemonicToRootKey s
+    }
+
+eitherToIO :: Either String a -> IO a
+eitherToIO (Right a) = return a
+eitherToIO (Left e) = hPutStrLn stderr e >> exitFailure
+
+byronCliKeyScheme :: ByronWalletStyle -> CliKeyScheme XPrv (Either String)
+byronCliKeyScheme = \case
+    Random ->
+        let
+            proxy = Proxy @'Random
+        in
+            CliKeyScheme
+                (allowedLengths proxy)
+                (fmap byronKeyFromSeed . seedFromMnemonic proxy)
+    Icarus ->
+        let
+            proxy = Proxy @'Icarus
+        in
+            CliKeyScheme
+                (allowedLengths proxy)
+                (fmap icarusKeyFromSeed . seedFromMnemonic proxy)
+    Trezor ->
+        let
+            proxy = Proxy @'Trezor
+        in
+            CliKeyScheme
+                (allowedLengths proxy)
+                (fmap hardwareKeyFromSeed . seedFromMnemonic proxy)
```

Comment:
Trezor /= Ledger. Trezor uses the same key generation method as Icarus. This is incidentally why the function is called generateKeyFromHardware**Ledger** !

Reply:
Ah!

Renaming `hardwareKeyFromSeed` to `ledgerKeyFromSeed` then. (And using `icarusKeyFromSeed` for Trezor.)